src/cli/telemetry/telemetry.ts

```diff
@@ -0,0 +1,304 @@
+import { tryFn, hasDebug, readJson, hasVerbose } from './helpers';
+import { shouldTrack } from './shouldTrack';
+import { CompilerBuildResults, Config, PackageJsonData, TaskCommand } from 'src/declarations';
+import { readConfig, updateConfig } from '../ionic-config';
+import { isObject } from '@utils';
+import { getCompilerSystem, getCoreCompiler, getStencilCLIConfig } from '../state/stencil-cli-config';
+
+/**
+ * Used as the object sent to the server. Value is the data tracked.
+ */
+export interface Metric {
+  name: string;
+  timestamp: string;
+  source: 'stencil_cli';
+  value: TrackableData;
+}
+
+/**
+ * Used for deeper understanding of how people are using certain output targets
+ **/
+interface OutputTargetOptions {
+  serviceWorker?: boolean;
+}
+
+/**
+ * The task to run in order to collect the duration data point.
+ */
+type TelemetryCallback = (...args: any[]) => void | Promise<void>;
+
+/**
+ * The model for the data that's tracked.
+ */
+export interface TrackableData {
+  yarn: boolean;
+  session_id: string;
+  component_count?: number;
+  arguments: string[];
+  targets: string[][] | OutputTargetOptions[][];
+  task: TaskCommand;
+  duration_ms: number;
+  packages: string[];
+  os_name: string;
+  os_version: string;
+  cpu_model: string;
+  typescript: string;
+  rollup: string;
+  system: string;
+  build: string;
+  stencil: string;
+}
+
+/**
+ * Used to within taskBuild to provide the component_count property.
+ * @param result The results of a compiler build.
+ * @returns void
+ */
+export async function telemetryBuildFinishedAction(result: CompilerBuildResults) {
+  const { sys, flags, logger, validatedConfig } = getStencilCLIConfig();
+  const tracking = await shouldTrack(!!flags?.ci);
+
+  if (!tracking) {
+    return;
+  }
+
+  const details = sys.details;
+  const session_id = await getTelemetryToken();
+  const packages = await getInstalledPackages();
+  const targets = await getActiveTargets(validatedConfig.config);
+  const versions = getCoreCompiler()?.versions || { typescript: 'unknown', rollup: 'unknown' };
+  const component_count = Object.keys(result.componentGraph).length;
+
+  const data: TrackableData = {
+    session_id,
+    yarn: isUsingYarn(),
+    duration_ms: result.duration,
+    component_count,
+    targets,
+    packages: packages || [],
+    arguments: flags.args,
+    task: flags.task,
```

Comment:
Is it possible that `flags` can be `undefined`? I thought that's why we used optional chaining [here](https://github.com/ionic-team/stencil/pull/2964/files#diff-2a84eb0343da5e51e060e0697c4c6fb67c2cc98b591ca94599ac984573787c3cR59). Maybe I'm missing something (likely)? Otherwise I'd be concerned the compiler is missing something there 🤔

Reply:
I've removed that optional chain.

Reply:
Does that solve the issue though? We define `_flags` as:
```typescript
export default class StencilCLIConfig {
  static instance: StencilCLIConfig;

  private _args: string[];
  private _logger: Logger;
  private _sys: CompilerSystem;
  private _flags: ConfigFlags | undefined;
  private _task: string | undefined;
  private _coreCompiler: CoreCompiler | undefined;
  private _validatedConfig: LoadConfigResults | undefined;

  private constructor(options: StencilCLIConfigArgs) {
    this._args = options?.args || [];
    this._logger = options?.logger;
    this._sys = options?.sys;
    this._flags = options?.flags || undefined;
    this._validatedConfig = options?.validatedConfig || undefined;
  }

  public static getInstance(options?: StencilCLIConfigArgs): StencilCLIConfig {
    if (!StencilCLIConfig.instance) {
      StencilCLIConfig.instance = new StencilCLIConfig(options);
    }

    return StencilCLIConfig.instance;
  }

  public get flags() {
    return this._flags;
  }
  public set flags(flags: ConfigFlags) {
    this._flags = flags;
  }
```
so it's possible we called the ctor for the first time with an empty object literal
```typescript
const baz = getInstance({});
```
which sets `_flags_` to `undefined` right? Barring setting that somewhere else, this would be an unsafe access

Reply:
Considering that use case you mentioned couldn't be possible as the code is written today, I don't think we'll experience that issue.